### PR TITLE
Add edit tree lemmatizer

### PR DESCRIPTION
**Pull request by adrianeboyd**
## Description

Add edit tree lemmatizer, converted from [`spacy_experimental.edit_tree_lemmatizer`](https://github.com/explosion/spacy-experimental/tree/d01fd5b479db823772865c362b4e4e1e706cf554/spacy_experimental/edit_tree_lemmatizer)

### Types of change

Enhancement

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [ ] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: spacy/pipeline/edit_trees.pxd
Comment (danieldk) [inline spacy/pipeline/edit_trees.pxd:3]: I think that these could now be relative imports?

File: spacy/tests/pipeline/test_edit_tree_lemmatizer.py
Comment (danieldk) [inline spacy/tests/pipeline/test_edit_tree_lemmatizer.py:11]: These could now be relative imports.

Comment (adrianeboyd) [reply spacy/tests/pipeline/test_edit_tree_lemmatizer.py:11]: The test suite intentionally uses absolute imports so it behaves the same when you run the tests on the installed package.

File: spacy/pipeline/edit_tree_lemmatizer.py
Comment (svlandeg) [reply spacy/pipeline/edit_tree_lemmatizer.py:141]: Sometimes you can also add in an `assert` if you're 100% of the type at that point in the code, and then `mypy` will use that in the subsequent lines.

Comment (adrianeboyd) [reply spacy/pipeline/edit_tree_lemmatizer.py:141]: Hmm, it doesn't like that either: 

`Parameterized generics cannot be used with class or instance checks`

Comment (adrianeboyd) [reply spacy/pipeline/edit_tree_lemmatizer.py:141]: The weird cast doesn't seem worse than ignoring it, but it seems like there ought to be a better solution.

Comment (danieldk) [reply spacy/pipeline/edit_tree_lemmatizer.py:141]: Maybe it would only work if the loss took `List[V]` where `V` is a typevar with `Ints1d` and `Floats2d` as the possible types. I am fine with keeping it as is.

Comment (danieldk) [reply spacy/pipeline/edit_tree_lemmatizer.py:141]: My only worry is that if we extend the union in the loss, type checking breaks here.

**Comment by adrianeboyd** [comment]
I'm not sure that `spacy.pipelines` was the best place for `edit_trees`, but there wasn't an immediately obvious better place to put it.

I'm also not sure that `edit_tree_lemmatizer` is the name we should go with in the end, since it's kind of opaque/jargon-y for users (I can't think of the actual name for this ... where the name is from the developers' perspective rather than the users').

My original intention was to have this kind of component as a `mode` of `lemmatizer`, which has tried to follow the `TrainablePipe` API enough to make it possible in the future (mainly having a `model` parameter), but now I think it will be an overall API headache to mix trainable and not-trainable modes.

**Comment by danieldk** [comment]
> I'm also not sure that `edit_tree_lemmatizer` is the name we should go with in the end, since it's kind of opaque/jargon-y for users (I can't think of the actual name for this ... where the name is from the developers' perspective rather than the users').

Maybe `TrainableLemmatizer`/`trainable_lemmatizer`?

**Comment by adrianeboyd** [comment]
So what I mean is the default component name / factory name and not the class name, where I think that `EditTreeLemmatizer` is fine.

Nothing else is `trainable_x` and "trainable" is expressed in other ways. But maybe there aren't any better alternatives...

File: spacy/pipeline/edit_tree_lemmatizer.py
Comment (svlandeg) [inline spacy/pipeline/edit_tree_lemmatizer.py:121]:
```suggestion
        self, examples: Iterable[Example], scores: List[Floats2d]
```

File: spacy/pipeline/edit_tree_lemmatizer.py
Comment (svlandeg) [inline spacy/pipeline/edit_tree_lemmatizer.py:221]:
```suggestion
    def label_data(self)  -> Dict:
```

File: spacy/pipeline/edit_tree_lemmatizer.py
Comment (svlandeg) [inline spacy/pipeline/edit_tree_lemmatizer.py:336]: Should we do a check here on `tree` being a properly structured dict with the correct fields? Or are these internals that can't (shouldn't) go wrong?

File: spacy/tests/pipeline/test_edit_tree_lemmatizer.py
Comment (svlandeg) [inline spacy/tests/pipeline/test_edit_tree_lemmatizer.py:189]: 🤩 😁

**Comment by svlandeg** [comment]
> I'm not sure that `spacy.pipelines` was the best place for `edit_trees`, but there wasn't an immediately obvious better place to put it.

I was thinking that too. But having `spacy.pipeline._edittree_internals` like we have for the `parser` is probably overkill. 

For the `entity_linker`, we did end up deciding having `kb.pyx` and `kb.pxd` in the root `spacy` module though.

**Comment by adrianeboyd** [comment]
Maybe `_internals` and we move the parser stuff there at some point, too?

**Comment by svlandeg** [comment]
> Nothing else is `trainable_x` and "trainable" is expressed in other ways. But maybe there aren't any better alternatives...

`trainable_x` does make the distinction more clear to users, I think, but I don't like it for the same reason you mention. Alternatively: `neural_lemmatizer`. Though it has the same issues, at least it's shorter :|

**Comment by svlandeg** [comment]
> Maybe `_internals` and we move the parser stuff there at some point, too?

I don't think that'll work, because the internal parser classes are so genericly named that then it might become more difficult to see what all the parser code is...

File: spacy/pipeline/edit_tree_lemmatizer.py
Comment (danieldk) [reply spacy/pipeline/edit_tree_lemmatizer.py:194]: I think I took this from the `Tagger` pipe, which does the same. But I agree that we should probably not do this.

Comment (danieldk) [reply spacy/pipeline/edit_tree_lemmatizer.py]: Fixed.

File: spacy/pipeline/edit_tree_lemmatizer.py
Comment (danieldk) [reply spacy/pipeline/edit_tree_lemmatizer.py]: Fixed.

File: spacy/pipeline/edit_tree_lemmatizer.py
Comment (danieldk) [reply spacy/pipeline/edit_tree_lemmatizer.py:332]: Since the trees can be passed through `initialized`, that would be good! I have written a schema using pydantic, but I am going to try to add this to `EditTrees.from_bytes` as well.

Comment (danieldk) [reply spacy/pipeline/edit_tree_lemmatizer.py]: I have added validation. I also had to change the on-disk format a bit, so that we can use the same schemas for both. While at it, I also simplified the on-disk format (it looks less like the C structs).

File: spacy/pipeline/edit_tree_lemmatizer.py
Comment (danieldk) [reply spacy/pipeline/edit_tree_lemmatizer.py:201]: I think my idea was to make it customizable what to do when no lemma could be predicted. However, if people want custom behavior, I think it is better that they disable backoff and add their own pipe to do post-processing.

Comment (danieldk) [reply spacy/pipeline/edit_tree_lemmatizer.py]: Removed the TODO comment.

**Review by svlandeg — APPROVED** [review]
Looks like this is ready to merge? 

We can revise the cast thing later, it'd require an update of Thinc and I don't want to hold up this PR over something like that.

**Comment by adrianeboyd** [comment]
No, not yet. I would wait a bit until the potential integrations with `Lemmatizer` have been explored a bit more.

**Comment by svlandeg** [comment]
Oh, ok. I thought we didn't want to go there initially, but I'll leave that up to you Adriane. I'll put this in draft to signal that we shouldn't merge yet!

**Comment by adrianeboyd** [comment]
We decided that an integration with `Lemmatizer` or the `lemmatizer` factory is not a particular useful choice. It's already supported to do something like this:

```
nlp.add_pipe("trainable_lemmatizer", name="lemmatizer")
```

**Comment by adrianeboyd** [comment]
Now this needs docs...

**Comment by danieldk** [comment]
> Now this needs docs...

Added, hope I didn't miss anything.

File: website/docs/usage/linguistic-features.md
Comment (adrianeboyd) [inline website/docs/usage/linguistic-features.md:206]:
```suggestion
transformations from a training corpus that includes lemma annotations. This removes the
```

File: website/docs/usage/processing-pipelines.md
Comment (adrianeboyd) [inline website/docs/usage/processing-pipelines.md:321]: Let's move this next to `lemmatizer`. (Or possible just sort this table, it is too long.)

File: website/docs/api/edittreelemmatizer.md
Comment (adrianeboyd) [inline website/docs/api/edittreelemmatizer.md:57]:
```suggestion
%%GITHUB_SPACY/spacy/pipeline/edit_tree_lemmatizer.pyx
```

File: website/docs/api/edittreelemmatizer.md
Comment (adrianeboyd) [inline website/docs/api/edittreelemmatizer.md:363]: Move to a yellow warning block?

File: website/docs/api/edittreelemmatizer.md
Comment (danieldk) [reply website/docs/api/edittreelemmatizer.md:57]: It's really `.py`, I changed it sometime when no Cython was needed in the lemmatizer itself.

File: website/docs/usage/linguistic-features.md
Comment (danieldk) [reply website/docs/usage/linguistic-features.md]: Fixed.

File: website/docs/usage/processing-pipelines.md
Comment (danieldk) [reply website/docs/usage/processing-pipelines.md]: Fixed.

File: website/docs/api/edittreelemmatizer.md
Comment (danieldk) [reply website/docs/api/edittreelemmatizer.md]: Yeah, that looks much better.

File: website/docs/api/edittreelemmatizer.md
Comment (danieldk) [reply website/docs/api/edittreelemmatizer.md]: Fixed.

**Pull request merged**
